### PR TITLE
feat: Add cloud profiler to training_utils

### LINKED ISSUE #519: Add remote tensorflow profiling to training jobs.
To support on demand profiling of tensorflow , add handlers for profiling requests when running training jobs.

- User updates their code to import the training util for profiling.
- User submits a training job to Vertex AI.
- Through Vertex AI Tensorboard, user can profile their remote training job through the "Capture Profile" functionality in Tensorboard.

### PR DESCRIPTION
**Pull request by mkovalski**
Adds ability to profile vertex training jobs using tensorboard profiler.

- Add a base plugin and tf profiler plugin to cloud training tools.
- Create helpers for uploading profiled items to tensorboard backend
- Add additional environment variables for setting webserver port.

Fixes #519

### CODE REVIEW

**Comment by nicain** [comment]
@mkovalski: I am assigning as owner of this PR; feel free to ping reviewers as needed to make sure the review process progresses in a timely fashion, or provide guidance on a who might better own the process of getting the PR reviewed, passing continuous testing, and merged. Reach out if you have questions.

File: setup.py [hunk 1]
```diff
@@ -36,8 +36,15 @@
 tensorboard_extra_require = ["tensorflow >=2.3.0, <=2.5.0"]
 metadata_extra_require = ["pandas >= 1.0.0"]
 xai_extra_require = ["tensorflow >=2.3.0, <=2.5.0"]
+profiler_extra_require = ["tensorboard-plugin-profile", "tensorflow >=2.4.0"]
+
 full_extra_require = list(
-    set(tensorboard_extra_require + metadata_extra_require + xai_extra_require)
+    set(
```
Comment (sasha-gitg) [inline setup.py:42]: TF version should be handled explicitly since TB, XAI, and Profiler have different version bounds.

**Review by sasha-gitg — CHANGES_REQUESTED** [review]

File: tests/unit/aiplatform/test_cloud_profiler.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2020 Google LLC
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply tests/unit/aiplatform/test_cloud_profiler.py:3]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/README.rst [hunk 3]
```diff
@@ -0,0 +1,20 @@
+Cloud Profiler
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/README.rst:1]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/initializer.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    prof_plugin = _build_plugin(plugin_obj)
+
+    if prof_plugin is None:
+        return
+
+    server = webserver.WebServer([prof_plugin])
+
+    if not environment_variables.http_handler_port:
+        raise MissingEnvironmentVariableException(
+            "'AIP_HTTP_HANDLER_PORT' must be set."
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/initializer.py:114]: This is set by the service.

File: google/cloud/aiplatform/training_utils/cloud_profiler/initializer.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    Args:
+        plugin (Callable[[], base_plugin):
+            Required. An uninitialized plugin.
+
+    Returns:
+        An initialized plugin, or None if plugin cannot be
+        initialized.
+    """
+    if not plugin.can_initialize():
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/initializer.py:52]: Yes Type is what I was going for and makes much more sense, thanks for this.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+from google.cloud.aiplatform.compat.types import (
+    tensorboard_experiment_v1beta1 as tensorboard_experiment,
+)
+from google.cloud.aiplatform.tensorboard.plugins.tf_profiler import profile_uploader
+from google.cloud.aiplatform import training_utils
+
+logger = tb_logging.get_logger()
+
+
+def _get_api_client():
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:42]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    region = m[1]
+
+    api_client = aiplatform.initializer.global_config.create_client(
+        client_class=TensorboardClientWithOverride, location_override=region,
+    )
+
+    return api_client
+
+
+def _get_project_id():
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:60]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+def _make_upload_limits():
+    upload_limits = server_info_pb2.UploadLimits()
+    upload_limits.min_blob_request_interval = 10
+    upload_limits.max_blob_request_size = 4 * (2 ** 20) - 256 * (2 ** 10)
+    upload_limits.max_blob_size = 10 * (2 ** 30)  # 10GiB
+
+    return upload_limits
+
+
+def _get_blob_items(api_client):
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:90]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    )
+    if not m:
+        raise ValueError(
+            "Incorrect format for tensorboard resource name: %s",
+            training_utils.environment_variables.tensorboard_resource_name,
+        )
+    return m[1]
+
+
+def _make_upload_limits():
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:81]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    path_prefix = tensorboard.blob_storage_path_prefix + "/"
+    first_slash_index = path_prefix.find("/")
+    bucket_name = path_prefix[:first_slash_index]
+    blob_storage_bucket = storage.Client(project=project_id).bucket(bucket_name)
+    blob_storage_folder = path_prefix[first_slash_index + 1 :]
+
+    return blob_storage_bucket, blob_storage_folder
+
+
+def _get_or_create_experiment(api, experiment_name):
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:105]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            training_utils.environment_variables.tensorboard_resource_name,
+            "experiments",
+            experiment_name,
+        )
+        experiment = api.get_tensorboard_experiment(name=experiment_name)
+
+    return experiment.name
+
+
+def create_profile_request_sender():
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:126]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+    if environment_variables.tf_profiler_port is None:
+        logger.warning(
+            '"%s" environment variable not set, cannot enable profiling.',
+            "AIP_TF_PROFILER_PORT",
+        )
+        return False
+
+    if environment_variables.tensorboard_log_dir is None:
+        logger.warning(
+            "Must set a tensorboard log directory, "
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:223]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+    def __init__(self):
+        """Build a TFProfiler object."""
+        from tensorboard_plugin_profile.profile_plugin import ProfilePlugin
+
+        context = _create_profiling_context()
+        self._profile_request_sender = tensorboard_api.create_profile_request_sender()
+        self._profile_plugin = ProfilePlugin(context)
+
+    def get_routes(self):
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:261]: Typing werkzeug environments feels a bit clunky, but I've added some type hints and corresponding PEP specifications. Let me know if you have any additional thoughts on this.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+        """List of routes to serve."""
+        return {"/capture_profile": self.capture_profile_wrapper}
+
+    # Define routes below
+    def capture_profile_wrapper(self, environ, start_response) -> Response:
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:266]: Typing werkzeug environments feels a bit clunky, but I've added some type hints and corresponding PEP specifications. Let me know if you have any additional thoughts on this.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+
+        self._profile_request_sender.send_request("")
+
+        return response
+
+    # End routes
+
+    @staticmethod
+    def setup() -> None:
+        import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:322]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+                        "invalid route for plugin %s, route %s"
+                        % (plugin.PLUGIN_NAME, route)
+                    )
+
+                app_route = os.path.join("/", plugin.PLUGIN_NAME)
+
+                app_route += route
+                self._routes[app_route] = handler
+
+    def dispatch_request(self, environ, start_response):
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py:61]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+        # Check for existince of route
+        request = wrappers.Request(environ)
+
+        if request.path in self._routes:
+            return self._routes[request.path](environ, start_response)
+
+        response = wrappers.Response("Not Found", status=404)
+        return response(environ, start_response)
+
+    def wsgi_app(self, environ, start_response):
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py:82]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        response = self.dispatch_request(environ, start_response)
+        return response
+
+    def __call__(self, environ, start_response):
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py:86]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/initializer.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+import threading
+from typing import Callable, Optional
+from werkzeug import serving
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/initializer.py:21]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/base_plugin.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import abc
+from typing import Callable, Dict
+from werkzeug import Response
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/base_plugin.py:20]: See https://github.com/googleapis/python-aiplatform/pull/828#discussion_r753313906

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Helpers for creating a profile request sender for tf profiler plugin."""
+
+import os
+import re
+
+from tensorboard.uploader import upload_tracker
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tensorboard_api.py:24]: See https://github.com/googleapis/python-aiplatform/pull/828#discussion_r753313906

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+#
+
+"""A plugin to handle remote tensoflow profiler sessions for Vertex AI."""
+
+import argparse
+from collections import namedtuple
+import importlib.util
+import json
+import logging
+from tensorboard.plugins.base_plugin import TBContext
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:25]: Done, see https://github.com/googleapis/python-aiplatform/pull/828#discussion_r753313906
Module level import updated.

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+
+
+class TFProfiler(base_plugin.BasePlugin):
+    """Handler for Tensorflow Profiling."""
+
+    PLUGIN_NAME = "profile"
+
+    def __init__(self):
+        """Build a TFProfiler object."""
+        from tensorboard_plugin_profile.profile_plugin import ProfilePlugin
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:272]: See https://github.com/googleapis/python-aiplatform/pull/828#discussion_r753313906

File: google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+
+import argparse
+from collections import namedtuple
+import importlib.util
+import json
+import logging
+from tensorboard.plugins.base_plugin import TBContext
+from typing import Optional
+from urllib import parse
+from werkzeug import Response
```
[the remark this answers is not in the document]
Comment (mkovalski) [reply google/cloud/aiplatform/training_utils/cloud_profiler/plugins/tensorflow/tf_profiler.py:28]: See https://github.com/googleapis/python-aiplatform/pull/828#discussion_r753313906

File: setup.py (hunk 1 above)
Comment (mkovalski) [reply setup.py]: Done.

File: google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sasha-gitg) [reply google/cloud/aiplatform/training_utils/cloud_profiler/webserver.py:102]: Was this method skipped on purpose?

**Review by sasha-gitg — APPROVED** [review]

**Pull request merged**